h5py/_objects.pyx

```diff
@@ -115,8 +116,19 @@ def nonlocal_close():
     """
     cdef ObjectID obj
 
-    # list() needed because the registry can be mutated concurrently
-    for python_id, ref in list(registry.items()):
+    # create a cached list of ids whilst the gc is disabled to avoid hitting
+    # the cyclic gc while iterating through the registry dict
+    gc.disable()
```

Comment:
This should be in a `try..finally` block.  Even though the `list` call _should_ be safe, we are here because it was not :smiling_imp: and because `KeyboardInterupt` can show up almost anywhere.